ASP.NET Core OData 8 pagination docs (#281)
MicrosoftDocs/OData-docs: Odata-docs/webapi-8/fundamentals/server-driven-paging.md

author: habbes
ms.author: clhabins
---

# Server-driven paging in ASP.NET Core OData 8
**Applies To**:[!INCLUDE[appliesto-webapi](../../includes/appliesto-webapi-v8.md)]

In [server-driven paging](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_ServerDrivenPaging), the server returns the first page of results. If total number of results is greater than the page size, the server returns the first page along with a "next link" that can be used to fetch the next page of results. Each subsequent page should include a next link except for the last page. Using this approach, the client needs to fetch pages sequentially. It cannot jump to an arbitrary page. The next link is included in the response using the [`@odata.nextLink`](http://docs.oasis-open.org/odata/odata-json-format/v4.01/odata-json-format-v4.01.html#sec_ControlInformationnextLinkodatanextL) annotation.

Here's an example of response:

```json
{
    "@odata.context": "http://localhost:5000/odata/$metadata#Products",
    "value": [
        {
            "Id": 1,
            "Name": "Product 1" 
        },
        {
            "Id": 2,
            "Name": "Product 2" 
        },
    ],

>     "@odata.nextLink": "http://localhost:5000/odata/Products?$skiptoken=foobar"

Reviewer: ```suggestion
    "@odata.nextLink": "http://localhost:5000/odata/Products?$skiptoken=Id-2"
```
To make this relatable with the actual value for skip token. `skiptoken=Id-2` simply means the last item in the returned result had an `Id` value of 2.

Author: This is an implementation detail. I wanted to show that the value of `$skiptoken` could be anything and the client shouldn't try to make sense of it.

}
```

In this example, the client would make a request to `GET http://localhost:5000/odata/Products?$skiptoken=foobar` to retrieve the next page.

> The format of the next link is opaque to the client. This means that the client should not try to decode the link, modify it or make any assumptions about how it is generated or what its components mean. It can only assume that it's a valid URL that will return the next page. This allows the server freedom in how it chooses to generate and interpret the next link. It allows the server to change the underlying implementation of the next link without affecting the client.

Reviewer 2: The client-driven paging doc explains why internally, `skip` isn't suitable but I was looking here for the explanation of how `skipToken` works internally so that it doesn't suffer the same issues.

Author: There's no standard way to handle the skip token, as this is up to the server's implementation. The default skip token handler in WebAPI 8 handles this conceptually similar to the example I showed using `GET /Products?$filter=Id gt 2&$top=2`. I talk more about the `DefaultSkipTokenHanlder` in a separate [PR](https://github.com/MicrosoftDocs/OData-docs/pull/283) that shows how to customize the `$skiptoken` logic. I can also briefly mention in this doc.

Author: I have added a brief explanation of how the default skip token handler handles the skip token and why the generated query is more efficient than using `$skip`.

## Configuring paging using `PageSize`

ASP.NET Core OData allow you to configure the page size on the server side using the [`PageSize`](/dotnet/api/microsoft.aspnetcore.odata.query.enablequeryattribute.pagesize) property of the [`[EnableQuery]`](/dotnet/api/microsoft.aspnetcore.odata.query.enablequeryattribute) attribute. This will automatically limit the number of items in the response and include an `@odata.nextLink` property.

The following code snippet demonstrates how to configure the `PageSize` on a controller action:

```c#
public class ProductsController : ODataController
{
    [EnableQuery(PageSize = 2)]
    public IQueryable<Product> Get()
    {
        return productsCollection;
    }
}
```

This sets the number of items per page to 2. Let's assume the entity set has a total of 5 products. To fetch the first page, simply fetch the entity set:

```http
GET http://localhost:5000/Products
```

**Response**:

```json
{
    "@odata.context": "http://localhost:5000/$metadata#Products",
    "value": [
        {
            "Id": 1,
            "Name": "Product 1"
        },
        {
            "Id": 2,
            "Name": "Product 2"
        }
    ],

>     "@odata.nextLink": "http://localhost:5000/Products?$skip=2"

Reviewer: ```suggestion
    "@odata.nextLink": "http://localhost:5000/Products?$skiptoken=Id-2"
```

Author: WebAPI uses `$skip` in the next link by default if you don't explicitly enable `SkipToken()`. I mention this in the "Improving pagint with `$skiptoken` section:

> You may have noticed that ASP.NET Core OData generates next links using `$skip`. Using `$skip` on large collections may lead to performance degradation in some databases. For more information about the drawbaks of `$skip`, visit the [client-driven paging](/odata/webapi-8/fundamentals/client-driven-paging#drawbacks) article.

}
```

The next link will fetch the second page:

```http
GET http://localhost:5000/Products?$skip=2
```

```json
{
    "@odata.context": "http://localhost:5000/$metadata#Products",
    "value": [
        {
            "Id": 3,
            "Name": "Product 3"
        },
        {
            "Id": 4,
            "Name": "Product 4"
        }
    ],

>     "@odata.nextLink": "http://localhost:5000/Products?$skip=4"

Reviewer: ```suggestion
    "@odata.nextLink": "http://localhost:5000/Products?$skiptoken=Id-4"
```

Author: See: https://github.com/MicrosoftDocs/OData-docs/pull/281#discussion_r1087433123

}
```

And finally, the last page:
```http
GET http://localhost:5000/Products?$skip=4
```
```json
{
    "@odata.context": "http://localhost:5000/$metadata#Products",
    "value": [
        {
            "Id": 3,
            "Name": "Product 3"
        }
    ]
}
```

The last page does not include a next link. This is how the client knows that it has reached the end of the collection.

If the request contains query options, they will also be included in the generated next link. Let's take the following request for example:

```http
GET http://localhost:5000/Products?$filter=Price gt 1000&$select=Name
```

The generated next link will look like:
```

> http://localhost:5000/Products?$filter=Price%20gt%201000&$select=Name&$skip=2

Reviewer: ```suggestion
http://localhost:5000/Products?$filter=Price%20gt%201000&$select=Name&$skiptoken=Id-2
```

Author: See: https://github.com/MicrosoftDocs/OData-docs/pull/281#discussion_r1087433123

```

## Combining `PageSize` and `$top`

`PageSize` automatically limits the number of items in the response. The client can still apply `$top` to the request. In this case, `$top` will limit the total number of results rather than the number of items per page.

If the client-requested `$top` is greater than `PageSize`, then the service should return the first page of results, with a next link that fetches the next page of results up to the maximum specified in the client's `$top`. If the `$top` is less that then page size, then server will return the number of items specified in the `$top` query option with no next link.

To illustrate this, let's use our entity set of 5 products as an example, with `PageSize` set to 2. The client makes a request with `$top=3`

```http
GET http://localhost:5000/Products?$top=3
```
**Response:**

```json
{
    "@odata.context": "http://localhost:5000/$metadata#Products",
    "value": [
        {
            "Id": 1,
            "Name": "Product 1"
        },
        {
            "Id": 2,
            "Name": "Product 2"
        }
    ],

>     "@odata.nextLink": "http://localhost:5000/Products?$top=1&$skip=2"

Reviewer: ```suggestion
    "@odata.nextLink": "http://localhost:5000/Products?$top=1&$skiptoken=Id-2"
```

Author: See: https://github.com/MicrosoftDocs/OData-docs/pull/281#discussion_r1087433123

}
```
The response contains 2 products and a next link to the next page. The client fetches the next page using the next link:

```http

> GET http://localhost:5000/Products?$top=1&$skip=2

Reviewer: ```suggestion
GET http://localhost:5000/Products?$top=1&$skiptoken=Id-2
```

Author: See: https://github.com/MicrosoftDocs/OData-docs/pull/281#discussion_r1087433123

```
**Response:**
```json
{
    "@odata.context": "http://localhost:5000/$metadata#Products",
    "value": [
        {
            "Id": 3,
            "Name": "Product 3"
        }
    ]
}
```
This is now the last page. The server returns only one item and no next link.

## Improving paging with `$skiptoken`

> You may have noticed that ASP.NET Core OData generates next links using `$skip`. Using `$skip` on large collections may lead to performance degradation in some databases. For more information about the drawbaks of `$skip`, visit the [client-driven paging](/odata/webapi-8/fundamentals/client-driven-paging#drawbacks) article.

Reviewer: I don't understand this paragraph. I expected that since this article talks about server-side paging, it'd focus on enabling skip token feature (`options.SkipToken()`) and applying `PageSize` configuration. My understanding was that those are the two cogs that drive server-side paging. However, what you're saying here is that `$skip` can be used to support server-side paging. That I wasn't aware. I might need to refresh my knowledge. I actually expected this article mostly on `$skiptoken` rather than `$skip`. It's also the reason why I left a lot of review suggestions to "fix" the next link

Author: The way I understand it, server-side paging means relying on the server-generated next link. The next link can be generated whether or not you use `$skiptoken`. The next link is computed when you set the `PageSize` property. Skip tokens provide an extra layer of improvements on top of next link. It's also something I discovered while writing this docq when I tried things out and referencing older docs we have on the subject:
- https://learn.microsoft.com/en-us/odata/webapi/pagesize-top-maxtop
- https://learn.microsoft.com/en-us/odata/webapi/skipToken-for-server-side-paging
- https://devblogs.microsoft.com/odata/using-skiptoken-for-paging-in-asp-net-odata-and-asp-net-core-odata/

I could focus only on `$skiptoken`, but I felt it would be incomplete. And given the common issues we get where customers confuse the roles of `PageSize` and `$top`, I thought it would be beneficial to dedicate a section on that. And given that `$skiptoken` is an improvement, I thought it would make more sense to "lead into it". I also considered that someone may come to this page without having read the client-driven paging docs. Open to suggestions.

Reviewer: @habbes Our offline conversation refers. It was a knowledge gap on my part. I'm happy with the document covering both `$skip` and `$skiptoken`

```c#
services.AddControllers().AddOData(options =>
    options.SetMaxTop(null).SkipToken());
```

Assuming we have the same products entity set as in the previous section and a `PageSize` of 2, here's what the first page would look like:

```http
GET http://localhost:5000/Products
```

**Response**

```json
{
    "@odata.context": "http://localhost:5000/$metadata#Products",
    "value": [
        {
            "Id": 1,
            "Name": "Product 1"
        },
        {
            "Id": 2,
            "Name": "Product 2"
        }
    ],
    "@odata.nextLink": "http://localhost:64771/Products?$skiptoken=Id-2"
}
```
The next link is now generated based on `$skiptoken` rather than `$skip`. This is what the second page would look like:

```
GET http://localhost:5000/Products?$skiptoken=Id-2
```

```json
{
    "@odata.context": "http://localhost:5000/$metadata#Products",
    "value": [
        {
            "Id": 3,
            "Name": "Product 3"
        },
        {
            "Id": 4,
            "Name": "Product 4"
        }
    ],
    "@odata.nextLink": "http://localhost:64771/Products?$skiptoken=Id-4"
}
```

> The `$skiptoken` value is generated using an implementation of [`SkipTokenHandler`](/dotnet/api/microsoft.aspnetcore.odata.query.skiptokenhandler). By default, the built-in [`DefaultSkipTokenHandler`](/dotnet/api/microsoft.aspnetcore.odata.query.defaultskiptokenhandler) class. `DefaultSkipTokenHandler` generates the `$skiptoken` based on the values of the key fields and fields in the `$orderby` query option if present.

Reviewer: ```suggestion
The `$skiptoken` value is generated using an implementation of [`SkipTokenHandler`](/dotnet/api/microsoft.aspnetcore.odata.query.skiptokenhandler), by default the built-in [`DefaultSkipTokenHandler`](/dotnet/api/microsoft.aspnetcore.odata.query.defaultskiptokenhandler) class. `DefaultSkipTokenHandler` generates the `$skiptoken` based on the values of the key fields and fields in the `$orderby` query option if present.
```

Author: Would that not make the sentence too long?

Reviewer: @habbes It's because "By default, the built-in DefaultSkipTokenHandler class." doesn't feel like a complete sentence

Reviewer 2: By default, the built-in [`DefaultSkipTokenHandler`](/dotnet/api/microsoft.aspnetcore.odata.query.defaultskiptokenhandler) class generates the `$skiptoken` based on the values of the key fields and fields in the `$orderby` query option if present

What about this?

Author: Aah I see, my bad.

Author: Rephrased to "By default, the built-in `DefaultSkipTokenHandler` class generates..."

> You can customize the `$skiptoken` by providing your own implementation of `SkipTokenHandler` and adding it to the OData route's service container:

Reviewer: ```suggestion
You can customize the `$skiptoken` by providing your own implementation of `SkipTokenHandler` and injecting it into the OData service dependency injection (DI) container:
```

---
title:  "Server-driven Paging in ASP.NET Core OData 8"
description: Server-driven paging in ASP.NET Core OData 8.
date:   2023-01-11
ms.date: 1/11/2023
author: habbes
ms.author: clhabins
---

# Server-driven Paging in ASP.NET Core OData 8

**Applies To**:[!INCLUDE[appliesto-webapi](../../includes/appliesto-webapi-v8.md)]

> In [server-driven paging](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_ServerDrivenPaging), the server returns the first page of results. If total number of results is greater than the page size, the server returns the first page along with a "next link" that can be used to fetch the next page of results. Each subsequent page should include a next link except for the last page. Using this approach, the client needs to fetch pages sequentially. It cannot jump to an arbitrary page. The next link is included in the response using the [`@odata.nextLink`](http://docs.oasis-open.org/odata/odata-json-format/v4.01/odata-json-format-v4.01.html#sec_ControlInformationnextLinkodatanextL) annotation.

Reviewer 3: you've provided a link to the v4.01 format but you are using a v4.0 format.

Author: Thanks. Updated.

Here's an example of response:

```json
{
    "@odata.context": "http://localhost:5000/odata/$metadata#Products",
    "value": [
        {
            "Id": 1,
            "Name": "Product 1" 
        },
        {
            "Id": 2,
            "Name": "Product 2" 
        },
    ],
    "@odata.nextLink": "http://localhost:5000/odata/Products?$skiptoken=foobar"
}
```

In this example, the client would make a request to `GET http://localhost:5000/odata/Products?$skiptoken=foobar` to retrieve the next page.

> The format of the next link is opaque to the client. This means that the client should not try to decode the link, modify it or make any assumptions about how it is generated or what its components mean. It can only assume that it's a valid URL that will return the next page. This allows the server freedom in how it chooses to generate and interpret the next link. It allows the server to change the underlying implementation of the next link without affecting the client.

Reviewer: ```suggestion
The format of the next link is opaque to the client. This means that the client should not try to decode the link, modify it or make any assumptions about how it is generated or what its components mean. It can only assume that it's a valid URL that will return the next page. This allows the server freedom in how it chooses to generate and interpret the next link. It also allows the server to change the underlying implementation of the next link without affecting the client.
```
